iceoryx_hoofs/include/iceoryx_hoofs/cxx/unique_ptr.hpp

```diff
@@ -40,10 +40,10 @@ class unique_ptr
     /// @brief unique_ptr Creates a unique pointer that takes ownership of an object.
     /// @details A deleter must always be provided as no default can be provided given that no heap is used.
     /// The unique_ptr must know how to delete the managed object when the pointer goes out of scope.
-    /// @param ptr The raw pointer to the object to be managed.
+    /// @param object The reference to the object to be managed.
     /// @param deleter The deleter function for cleaning up the managed object.
     ///
-    unique_ptr(T* const ptr, const function<void(T*)>& deleter) noexcept;
+    unique_ptr(T& object, const function<void(T*)>& deleter) noexcept;
```

Comment:
Please not. This is not idiomatic C++. A reference does not transfer ownership but the class takes the ownership of the pointer/reference. I know what you try to do but this is the wrong way to do it. Can you please use `cxx::non_null`? Maybe we can extend it to have a ctor which takes a reference.

Reply:
Here I would disagree, especially since it is idiomatic C++ in my opinion. Stating the C++ FAQ from isocpp.org: 

```
Use references when you can, and pointers when you have to.

References are usually preferred over pointers whenever you don’t need [“reseating”]. This usually means that references are most useful in a class’s public interface. References typically appear on the skin of an object, and pointers on the inside.
```

see: https://isocpp.org/wiki/faq/references#refs-vs-ptrs

Furthermore it states at the end:
```
Note: Old line C programmers sometimes don’t like references since they provide reference semantics that isn’t explicit in the caller’s code. After some C++ experience, however, one quickly realizes this is a form of information hiding, which is an asset rather than a liability. E.g., programmers should write code in the language of the problem rather than the language of the machine.
```

So I suspect that you are an old line c programmer :smile:

Reply:
I disagree. The intention of this rule is to pass references instead of raw pointer when you have an object.
```cpp
Foo foo;
bar(&foo); // bad
bar(foo) // good
```
It is different if you already have a pointer, which is the case when you use a `unique_ptr`.
```cpp
Foo* foo {nullptr};
bar(foo); // good when the nullptr check is done in bar
bar(*foo); // bad since we have a null reference
```

But it seems we are both wrong [I.11: Never transfer ownership by a raw pointer (T*) or reference (T&)](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#i11-never-transfer-ownership-by-a-raw-pointer-t-or-reference-t) :laughing: 
That link refers to returning a pointer not to create a smart pointer. I would argue that it is more idiomatic to pass a raw pointer to as constructor of a smart pointer than a reference. Which leads us to [I.12: Declare a pointer that must not be null as not_null](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#i12-declare-a-pointer-that-must-not-be-null-as-not_null)

```cpp
Foo* flup {new Foo()}; 
bar(non_null<T>(flup)); // better
bar(make_non_null(flup).value()) // make_non_null would return a `optional<non_null<T>>`
bar(make_non_null_unsafe_unchecked(flup)); // for if performance matters
```